pkg/dnsmasq/patches/lease_release.patch

```diff
@@ -0,0 +1,25 @@
+diff -r -u ../orig/dnsmasq-2.78/src/rfc2131.c src/rfc2131.c
+--- a/src/rfc2131.c 2017-10-02 18:43:51.000000000 +0530
++++ b/src/rfc2131.c	2020-04-09 15:31:15.000000000 +0530
+@@ -1026,11 +1026,10 @@
+ 		  int len;
+ 		  unsigned char *mac = extended_hwaddr(ltmp->hwaddr_type, ltmp->hwaddr_len,
+ 						       ltmp->hwaddr, ltmp->clid_len, ltmp->clid, &len);
+-		  my_syslog(MS_DHCP | LOG_WARNING, _("not using configured address %s because it is leased to %s"),
++		  my_syslog(MS_DHCP | LOG_WARNING, _("pruning configured address %s because it is leased to %s"),
+ 			    addrs, print_mac(daemon->namebuff, mac, len));
++		  lease_prune(ltmp, now);
```

Comment:
Does this prune actually cause the lease to happen? Or does it depend on a retransmit of the DHCP request from the client?

Is this something which would be usable for others, so that we can try to upstream it?
I wonder when the config in the dhcp-hosts file/directory conflicts which an existing lease, whether that is a general issue or just something we run in to because we change things frequently.

Reply:
This patch prunes the existing entry in dhcpcd.leases and creates/responds with new entry.
Might not be very useful for others. The case we hit happens when we statically allocate the same IP address for two different mac addresses. Most might just make dnsmasq allocate IP address for this use case.

Reply:
Yeah and ideally we should have the domU release the lease before it goes down, but that might not be common. Especially not for a container.

The other thing we can do to reduce the incidence of this would be to not immediately reuse an appID after it has been freed, so that we are likely to get a less recently used MAC address for a new app instance.

I don't know what is the best apprioach here.

Reply:
Actually, I don't think the old app id is being used in this case. It is because a new app id is being used, we create a new mac address. But the IPAM code starts probing the allocation vector from beginning. So, we end up allocating same IP address for both old and new mac addresses. This leads to the clash in leases.

Reply:
If appId (well, appNum - sorry for the confusion) is part of the mac address and it can be reused immediately after the previous app instance is deleted.If we instead have appnumallocator try the next free number (and round-robin once it gets to 255) then the MAC addresses would not be reused that quickly and dnsmasq shouldn't have a problem.

Reply:
But, in this case app num is changing and there by the mac address between old/new app instance. It is the allocated IP address that does not change.

Reply:
@gkodali-zededa is the failure due to quickly reusing the mac address and expecting to get a different IP address (due to a different network instance etc). Or is it due to wanting to reuse the IP address for a different app instance?

Reply:
This is what happened @eriknordmark :
1) ubuntu app was started with eth0. Zedrouter gave eth0 a mac address of xx:xx:xx:xx:xx:01 and IP address 172.168.1.2
2) dnsmasq added this map from xx:xx:xx:xx:xx:01 -> 172.168.1.2 and lease time 10 minutes.
3) Then the ubuntu app was deleted. As a result zedrouter freed the IP address 172.168.1.2
4) The the same app was added back again. Zedrouter now gave eth0 the mac xx:xx:xx:xx:xx:02 and IP same as before (172.168.1.2).
5) dnsmasq complained that this particular IP address has already been given to xx:xx:xx:xx:xx:01 and not reply to DHCP request from app.
6) After the 10 miutes elapsed, lease expired and the app got it's IP address.

Reply:
OK, so then changing the appNum allocation would do nothing. Pulling this PR.